polaris-core/src/main/java/org/apache/polaris/core/identity/ServiceIdentityType.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.core.identity;
+
+import jakarta.annotation.Nonnull;
+import java.util.Arrays;
+import org.apache.polaris.core.identity.dpo.ServiceIdentityInfoDpo;
+
+/**
+ * The internal persistence-object counterpart to ServiceIdentityTypeInfo.ServiceIdentityTypeEnum
+ * defined in the API model. We define integer type codes in this enum for better compatibility
+ * within persisted data in case the names of enum types are ever changed in place.
+ *
+ * <p>Important: Codes must be kept in-sync with JsonSubTypes annotated within {@link
+ * ServiceIdentityInfoDpo}.
+ */
+public enum ServiceIdentityType {
+  NULL_TYPE(0),
+  AWS_IAM(1),
+  ;
+
+  private static final ServiceIdentityType[] REVERSE_MAPPING_ARRAY;
+
+  static {
+    // find max array size
+    int maxCode =
+        Arrays.stream(ServiceIdentityType.values())
+            .mapToInt(ServiceIdentityType::getCode)
+            .max()
+            .orElse(0);
+
+    // allocate mapping array
+    REVERSE_MAPPING_ARRAY = new ServiceIdentityType[maxCode + 1];
+
+    // populate mapping array
+    for (ServiceIdentityType authType : ServiceIdentityType.values()) {
+      REVERSE_MAPPING_ARRAY[authType.code] = authType;
+    }
+  }
+
+  private final int code;
+
+  ServiceIdentityType(int code) {
+    this.code = code;
+  }
+
+  /**
+   * Given the code associated with the type, return the associated ServiceIdentityType. Return
+   * NULL_TYPE if not found
+   *
+   * @param identityTypeCode code associated to the service identity type
+   * @return ServiceIdentityType corresponding to that code or null if mapping not found
+   */
+  public static @Nonnull ServiceIdentityType fromCode(int identityTypeCode) {
+    // ensure it is within bounds
+    if (identityTypeCode < 0 || identityTypeCode >= REVERSE_MAPPING_ARRAY.length) {
+      return NULL_TYPE;
```

Comment:
What is the purpose of `NULL_TYPE`? It has a specific type code (`0`) but we map to it from other specific type codes... how is that meant to work end-to-end? :thinking:

Reply:
It's just a enum value for all the out-of-bounds codes

With this, we don't need to check `serviceIdentityType != null` since it will always return an nonnull value!

Reply:
I do not see any Polaris code where `serviceIdentityType` is checked at all :)

Reply:
My point in the original method is that if the idea is to keep working without error, then we're losing information in `fromCode()`. What the benefit of returning a non-null value, when the original value's meaning is lost?

Reply:
Right now, ServiceIdentityType is not used. But In part 3, ServiceIdentityInfo will be assigned and injected to catalog entity and then serviceIdentityType is not null and will be used for the further identity resolution (resolve the identity, get the credential and then make a remote call)

Reply:
To put it another way: why bother with an enum in this case when the type code is an `int` and code's behaviour appears to be based on `ServiceIdentityInfoDpo` sub-types' details ?

Reply:
Just to make it slightly easier for downstream implementations, without needing to check whether the returned value is null. But yes, the original type code's meaning is lost.

But we should be able to original code if the returned type is `NULL` right, since the caller passes the code to this method

Reply:
This  enum is a bridge between serialized (int) and runtime (class) representation (we don't want to compare the int with 1/2/3, the code is very confusing and hard to read).

We will use this `ServiceIdentityType` in both dpo and the runtime `ResolvedServiceIdentity` class (will be added in part 3).